### PR TITLE
Rework concurrency on `ManifestStorage`

### PR DESCRIPTION
**Pull request by FirelightFlagboy**

### CODE REVIEW

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 1]
```diff
@@ -92,16 +95,29 @@ pub struct ManifestStorage {
     local_symkey: SecretKey,
     conn: Arc<Mutex<SqliteConn>>,
     pub realm_id: EntryID,
+    pub(crate) caches: Arc<Mutex<HashMap<EntryID, Arc<Mutex<Cache>>>>>,
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:98]: Why do you need Arc for the values ?

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:98]: The goal is to not lock the entire map, when manipulating a single value.

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 2]
```diff
@@ -379,12 +396,22 @@ impl ManifestStorage {
         Ok((local_changes, remote_changes))
     }
 
+    pub(crate) fn get_cache_entry(&self, entry_id: EntryID) -> Arc<Mutex<CacheEntry>> {
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:399]:
```suggestion
    pub(crate) fn get_or_insert_default_cache_entry(&self, entry_id: EntryID) -> Arc<Mutex<CacheEntry>> {
```
also add a doc-comment

**Review by vxgmichel — APPROVED** [review]
Looks neat!

File: oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs [hunk 3]
```diff
@@ -94,14 +94,26 @@ pub struct ManifestStorage {
...
-    /// the chunks that needs to be removed from the conn after the
-    /// manifest is written. Note: this set might be empty but the manifest
-    /// still requires to be flushed.
-    cache_ahead_of_localdb: Arc<Mutex<HashMap<EntryID, HashSet<ChunkOrBlockID>>>>,
+    caches: Arc<Mutex<HashMap<EntryID, Arc<Mutex<CacheEntry>>>>>,
+}
+
+/// A cache entry that may contain a manifest and its pending chunk.
+#[derive(Default)]
+pub(crate) struct CacheEntry {
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/core_fs/src/storage/manifest_storage.rs:102]: I've tried it but we rapidly hit ownership problems when we tried to update a variant to another without cloning everywhere.

**Pull request merged**